packages/core/src/api/serialization/serialization.ts

```diff
@@ -0,0 +1,182 @@
+// TODO: IMO this should be part of the formatConversions file since the custom
+//  serializer is used for all HTML & markdown conversions. I think this would
+//  also clean up testing converting to clean HTML.
+import { Extension } from "@tiptap/core";
+import { Plugin } from "prosemirror-state";
+import { DOMSerializer, Fragment, Node, Schema } from "prosemirror-model";
+import { nodeToBlock } from "../nodeConversions/nodeConversions";
+import { BlockNoteEditor } from "../../BlockNoteEditor";
+import {
+  BlockSchema,
+  SpecificBlock,
+} from "../../extensions/Blocks/api/blockTypes";
+import { cleanHTML, markdown } from "../formatConversions/formatConversions";
+
+const acceptedMIMETypes = [
+  "blocknote/html",
+  "text/html",
+  "text/plain",
+] as const;
+
+function doc(options: { document?: Document }) {
+  return options.document || window.document;
+}
+
+export const customBlockSerializer = <BSchema extends BlockSchema>(
+  schema: Schema,
+  editor: BlockNoteEditor<BSchema>
+) => {
+  const customSerializer = DOMSerializer.fromSchema(schema) as DOMSerializer & {
+    serializeNodeInner: (
+      node: Node,
+      options: { document?: Document }
+    ) => HTMLElement;
+  };
+
+  customSerializer.serializeNodeInner = (
+    node: Node,
+    options: { document?: Document }
+  ) => {
+    const { dom, contentDOM } = DOMSerializer.renderSpec(
+      doc(options),
+      customSerializer.nodes[node.type.name](node)
+    );
+
+    if (contentDOM) {
+      if (node.isLeaf) {
+        throw new RangeError("Content hole not allowed in a leaf node spec");
+      }
+
+      // Checks if the block type is custom. Custom blocks don't implement a
+      // `renderHTML` function in their TipTap node type, so `toDOM` also isn't
+      // implemented in their ProseMirror node type.
+      if (
+        node.type.name === "blockContainer" &&
+        node.firstChild!.type.spec.toDOM === undefined
+      ) {
+        // Renders block content using the custom `blockSpec`'s `serialize`
+        // function.
+        const blockContent = DOMSerializer.renderSpec(
+          doc(options),
+          editor.schema[node.firstChild!.type.name as keyof BSchema].serialize!(
+            nodeToBlock(
+              node,
+              editor.schema,
+              editor.blockCache
+            ) as SpecificBlock<BlockSchema, string>,
```

Comment:
I think the typing for `nodeToBlock` can just be changed to return a `SpecificBlock`. We'll also need to cast stuff to do this but it's kinda inevitable since PM nodes aren't strongly typed.

Reply:
I made an effort to improve this but the casting is still pretty nasty. The underlying issue is that when we convert a ProseMirror `Node` to a `Block`, we have basically 0 type information. I managed to work around this issue to some extent for the custom blocks API with the `TipTapNode` type and `createTipTapBlock` function, but this only works because we're creating the node in the first place.